cloudinit/sources/DataSourceAzure.py

```diff
@@ -445,19 +444,30 @@ def crawl_metadata(self):
                 else:
                     ret = load_azure_ds_dir(cdev)
 
-            except NonAzureDataSource:
-                report_diagnostic_event(
-                    "Did not find Azure data source in %s" % cdev,
-                    logger_func=LOG.debug)
+            except (NonAzureDataSource, util.MountFailedError) as e:
```

Comment:
This for loop feels increasingly overloaded to me: it feels like we have some unnecessary indirection by putting "magic" values into the list, when it would probably be simpler to circumvent the list entirely. Would something like this pseudocode make sense (as a replacement from ~line 411):

```py
ret = None
if os.path.isfile(REPROVISION_MARKER_FILE):
    ret, reprovision = None, True
elif os.path.isfile(NIC_ATTACH_MARKER_PRESENT):
    ret, reprovision_after_nic_attach = None, True
else:
     try:
         ret = load_azure_ds_dir(self.seed_dir)
    except (...):
        for cdev in list_possible_azure_ds_devs():
            try:
                ret = util.mount_cb(...)
                break
            except (...):
                ...
                continue
    if not ret:
        ret = load_azure_ds_dir(ddir)
```

You may point out that this retains the previous behaviour, rather than the behaviour you're changing to. I'd suggest that we can address that by converting `list_possible_azure_ds_devs()` to a generator; very roughly:

```py
def list_possible_azure_ds_devs(...):
    yield DEFAULT_PROVISIONING_ISO_DEV
    for fstype in ("iso9660", "udf"):
        yield from util.find_devs_with("TYPE=%s" % fstype)
```

This will yield `DEFAULT_PROVISIONING_ISO_DEV` first, and only if a second value is requested will it execute `util.find_devs_with(...)`: if `DEFAULT_PROVISIONING_ISO_DEV` works, then we'll `break` out of the for loop before . (This would actually be an improvement even without the short-circuiting for this PR: it will only call `util_find_devs_with` for UDF if none of the ISO9660 options work.)

Additionally: I think that everything within the existing for loop between the exception handlers and `break` could move out of the for loop: it will be executed at most once anyway, due to the `break`: it doesn't need to go inside my proposed for loop here.

What do you think?

Reply:
Thanks for the suggestion on switching list_possible_azure_ds_devs to a generator. I've made the changes